File: multi_language_code_generator/reference/dependency_resolver.py
from collections import deque, defaultdict

class DependencyResolver:
    def resolve(self, dependencies: dict) -> dict:
        """
        Resolves the build order for a given dictionary of dependencies using
        Kahn's Algorithm for Topological Sorting.
        
        Args:
            dependencies: Dict where key is package name and value is list of dependencies.
        
        Returns:
            Dict with 'success', 'build_order', and 'error'.
        """
        # 1. Normalize graph and calculate in-degrees
        in_degree = defaultdict(int)
        graph = defaultdict(list)
        all_nodes = set(dependencies.keys())

        # Ensure all nodes mentioned in values are also tracked
        for pkg, deps in dependencies.items():
            for dep in deps:
                all_nodes.add(dep)

        # Build the graph (dependency -> dependent) because if A depends on B,
        # B must come before A. So the edge is B -> A.
        for pkg, deps in dependencies.items():
            for dep in deps:
                graph[dep].append(pkg)
                in_degree[pkg] += 1
        
        # 2. Initialize queue with nodes having 0 in-degree (no dependencies)
        queue = deque([node for node in all_nodes if in_degree[node] == 0])
        sorted_order = []

        # 3. Process queue
        while queue:
            # Sort for deterministic output when multiple nodes are ready simultaneously
            # (Note: In a real large-scale scenario, sorting the queue is inefficient, 
            # but valid for stability in this reference implementation)
            current = queue.popleft() 
            sorted_order.append(current)

            if current in graph:
                for neighbor in graph[current]:
                    in_degree[neighbor] -= 1
                    if in_degree[neighbor] == 0:
                        queue.append(neighbor)
        
        # 4. Check for cycles
        if len(sorted_order) != len(all_nodes):
            return {
                "success": False,
                "build_order": None,
                "error": "Cycle detected or unresolvable dependencies."
            }

        return {
            "success": True,
            "build_order": sorted_order,
            "error": None
        }

File: multi_language_code_generator/implementations/javascript/dependencyResolver.js
class DependencyResolver {
    /**
     * Resolves the build order using topological sort.
     * @param {Object} dependencies - Key: package name, Value: array of dependencies.
     * @returns {Object} - { success, build_order, error }
     */
    resolve(dependencies) {
        const inDegree = new Map();
        const graph = new Map();
        const allNodes = new Set(Object.keys(dependencies));

        // 1. Normalize: ensure all referenced deps are in allNodes
        for (const deps of Object.values(dependencies)) {
            deps.forEach(dep => allNodes.add(dep));
        }

        // Initialize data structures
        allNodes.forEach(node => {
            inDegree.set(node, 0);
            graph.set(node, []);
        });

        // 2. Build Graph (Dependency -> Dependent)
        // If A depends on B, B must precede A. Edge B -> A.
        for (const [pkg, deps] of Object.entries(dependencies)) {
            deps.forEach(dep => {
                graph.get(dep).push(pkg);
                inDegree.set(pkg, inDegree.get(pkg) + 1);
            });
        }

        // 3. Initialize Queue (using Array + sort for determinism)
        let queue = [];
        allNodes.forEach(node => {
            if (inDegree.get(node) === 0) {
                queue.push(node);
            }
        });
        
        // Sort initial queue for consistency with Python/Java
        queue.sort();

        const sortedOrder = [];

        // 4. Process Queue
        while (queue.length > 0) {
            // Simulate queue behavior: remove from front
            const current = queue.shift();
            sortedOrder.push(current);

            if (graph.has(current)) {
                const neighbors = graph.get(current);
                // Sort neighbors for consistency
                neighbors.sort();
                
                neighbors.forEach(neighbor => {
                    inDegree.set(neighbor, inDegree.get(neighbor) - 1);
                    if (inDegree.get(neighbor) === 0) {
                        queue.push(neighbor);
                    }
                });
                
                // After pushing new nodes, we must re-sort if we want strict
                // lexicographical processing order across the whole run,
                // but standard Kahn's just adds to end. 
                // To match the Python/Java PriorityQueue logic exactly:
                // We re-sort the queue after every insertion or just allow 
                // the natural "level" order. For strict parity with the Java 
                // PriorityQueue implementation above, we resort:
                queue.sort();
            }
        }

        // 5. Check for Cycles
        if (sortedOrder.length !== allNodes.size) {
            return {
                success: false,
                build_order: null,
                error: "Cycle detected or unresolvable dependencies."
            };
        }

        return {
            success: true,
            build_order: sortedOrder,
            error: null
        };
    }
}

module.exports = DependencyResolver;
